lib/clientv2/ippool.go

```diff
@@ -145,22 +210,58 @@ func (_ ipPools) validateAndSetDefaults(pool *apiv2.IPPool) error {
 		}
 	}
 
+	// Normalize the CIDR before persisting.
+	new.Spec.CIDR = cidr.String()
+
+	// If there was a previous pool then this must be an Update, validate that the
+	// CIDR has not changed.  Since we are using noramlized CIDRs we can just do a
+	// simple string comparison.
+	if old != nil && old.Spec.CIDR != new.Spec.CIDR {
+		errFields = append(errFields, cerrors.ErroredField{
+			Name:   "IPPool.Spec.CIDR",
+			Reason: "IPPool CIDR can not be modified",
+		})
+	}
+
+	// If there was no previous pool then this must be a Create.  Check that the CIDR
+	// does not overlap with any other pool CIDRs.
+	if old == nil {
+		allPools, err := r.List(ctx, options.ListOptions{})
+		if err != nil {
+			return err
+		}
+
+		for _, otherPool := range allPools.Items {
+			_, otherCIDR, err := cnet.ParseCIDR(otherPool.Spec.CIDR)
+			if err != nil {
+				log.WithField("Name", otherPool.Name).WithError(err).Error("IPPool is configured with an invalid CIDR")
+				continue
+			}
+			if otherCIDR.IsNetOverlap(cidr.IPNet) {
+				errFields = append(errFields, cerrors.ErroredField{
+					Name:   "IPPool.Spec.CIDR",
+					Reason: fmt.Sprintf("IPPool CIDR overlaps with IPPool(%s) CIDR %s", otherPool.Name, otherPool.Spec.CIDR),
```

Comment:
Note: this causes an existing calicoctl ST to fail:
```
FAIL: Test namespace is handled as expected for each resource type [with data={'kind': 'IPPool', 'spec': {'ipi...data': {'name': 'ippool-name1'}}, namespaced=False].
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 392, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/code/tests/st/calicoctl/test_crud.py", line 364, in test_namespace
    rc.assert_no_error()
  File "/code/tests/st/utils/utils.py", line 139, in assert_no_error
    "command=" + self.command + "\noutput=" + self.output
AssertionError: Expected no error running command; 
command=export ETCD_ENDPOINTS=http://192.168.1.16:2379; export ETCD_CA_CERT_FILE=; export ETCD_CERT_FILE=; export ETCD_KEY_FILE=; export DATASTORE_TYPE=etcdv3;  /code/dist/calicoctl create -f /tmp/input-data
output=Failed to create 'IPPool' resource: error with field IPPool.Spec.CIDR = '<nil>' (IPPool CIDR overlaps with IPPool(name1) CIDR 10.0.1.0/24)
```

(https://semaphoreci.com/calico/calicoctl/branches/pull-request-1742/builds/1)

Reply:
If this is an intended restriction now (which I think it is!), it should be documented at https://docs.projectcalico.org/master/reference/calicoctl/resources/ippool.  (As well as fixing up the ST.)

Reply:
d'ohhhhhhh   - silly me!

thanks!  I'll jump on this.

1. is because none of the validation, including the stuff I've just added includes the field value.

I think there are a few fixes required here.  Firstly to add the values to the validation errors, secondly for the validation error to handle more gracefully a value not being present - I don't think we want to display nil values in the error messages.